@xen-orchestra/rest-api/src/vms/vm.controller.mts

```diff
@@ -205,4 +207,42 @@ export class VmController extends XapiXoController<XoVm> {
       },
     })
   }
+
+  /**
+   * @example id "f07ab729-c0e8-721c-45ec-f11276377030"
+   * @example body { "name_label": "my_awesome_snapshot" }
+   */
+  @Example(taskLocation)
+  @Post('{id}/actions/snapshot')
+  @SuccessResponse(actionAsyncroneResp.status, actionAsyncroneResp.description, actionAsyncroneResp.produce)
+  @Response(createdResp.status, 'Snapshot created')
+  @Response(notFoundResp.status, notFoundResp.description)
+  @Response(internalServerErrorResp.status, internalServerErrorResp.description)
+  async snapshotVm(
+    @Path() id: string,
+    @Body() body?: { name_label?: XoVmSnapshot['name_label'] },
+    @Query() sync?: boolean
+  ): Promise<string | { uuid: XenApiVm['uuid'] }> {
+    const vmId = id as XoVm['id']
+    const action = async () => {
+      const xapiVm = this.getXapiObject(vmId)
+      const ref = await xapiVm.$snapshot({ ignoredVdisTag: '[NOSNAP]', name_label: body?.name_label })
+      const snapshotUuid = await xapiVm.$xapi.getField<XenApiVm, 'uuid'>('VM', ref, 'uuid')
+
+      if (sync) {
+        this.setHeader('Location', `${BASE_URL}/vm-snapshots/${snapshotUuid}`)
+      }
```

Comment:
I think the create action overwrite the header if needed. 
Why do this action try to do a redirection ? shouldn't it return the snapshot uuid ?

Reply:
`createAction` writes `setHeader` in case of `sync` false. Besides, I think the `createAction` method should handle itself if it needs to write a `header` location in case of a `sync` action (for example, if the status code is `201`).

This is not a redirect, but simply information in the response header.

Reply:
The action returns the snapshot's UUID:
`return {uuid:snapshotUuid}`.
`this.setHeader('Location)` simply adds the "where to find the created resource" information to the response header in case of sync request